src/main/java/com/aws/greengrass/shadowmanager/PubSubIntegrator.java

```diff
@@ -0,0 +1,173 @@
+        this.updateThingShadowRequestHandler = updateThingShadowRequestHandler;
+        this.getThingShadowRequestHandler = getThingShadowRequestHandler;
+    }
+
+    /**
+     * Subscribes to the shadow topic over local PubSub.
+     */
+    public void subscribe() {
+        if (this.subscribed.compareAndSet(false, true)) {
+            this.pubSubClientWrapper.subscribe(this::handlePublishedMessage);
+        }
+    }
+
+    /**
+     * Unsubscribes to the shadow topic over local PubSub.
+     */
+    public void unsubscribe() {
+        if (this.subscribed.compareAndSet(true, false)) {
+            this.pubSubClientWrapper.unsubscribe(this::handlePublishedMessage);
+        }
+
+    }
+
+    /**
+     * Helper function to extract the thingName, shadowName and shadow operation from mqtt topic.
+     *
+     * @param topic PubSub topic on which the message was sent.
+     * @return ShadowRequest object with shadow details
+     */
+    ShadowRequest extractShadowFromTopic(String topic) {
+        final Matcher matcher = shadowPattern.matcher(topic);
+
+        if (matcher.find() && matcher.groupCount() == 4) {
+            String thingName = matcher.group(1);
+            String shadowName = matcher.group(3);
+            String operation = matcher.group(4);
+            return new ShadowRequest(thingName, shadowName, operation);
+        }
+        logger.atWarn()
+                .kv("topic", topic)
+                .log("Unable to parse shadow topic for thing name, shadow name and shadow operation");
+        throw new IllegalArgumentException("Unable to parse shadow topic for thing name shadow name "
+                + "and shadow operation");
+    }
+
+    /**
+     * Handle the new message that is published oveg local PubSub. Extract the necessary information from the shadow
+     * topic to accurately route the message to the appropriate handler.
+     * It will ignore the message if it is unable to extract all the necessary information from the shadow topic.
+     *
+     * @param publishEvent The message that is published over local PubSub.
+     */
+    private void handlePublishedMessage(PublishEvent publishEvent) {
+        String topic = publishEvent.getTopic();
+        logger.atDebug().kv(LOG_TOPIC, topic).log("Processing new shadow operation message over local PubSub");
+        try {
+            ShadowRequest shadowRequest = extractShadowFromTopic(topic);
+            switch (shadowRequest.getOperation().toLowerCase()) {
+                case "update":
+                    UpdateThingShadowRequest request = new UpdateThingShadowRequest();
+                    request.setThingName(shadowRequest.getThingName());
+                    request.setShadowName(shadowRequest.getShadowName());
+                    request.setPayload(publishEvent.getPayload());
+
+                    this.updateThingShadowRequestHandler.handleRequest(request, SHADOW_MANAGER_NAME);
```

Comment:
what happens on a ServiceError - does that bubble up and just get logged?

Can you add some test cases for the handler throwing an exception?